출저 : 주니온TV(유튜브)

In [ ]:
ㅇㅇㅇㅇㅇㅇㅇㅇ

## 콜라츠의 추측: 우박수 문제

 - 쉬운데 어렵다. 대학에서 잘 가르킨다.
 ![image-20210927120303710](https://raw.githubusercontent.com/is3js/screenshots/main/image-20210927120303710.png)

 ![image-20210927120351739](https://raw.githubusercontent.com/is3js/screenshots/main/image-20210927120351739.png)


 

### 구현(collatz) : 콜라츠 수열을 만들어 반환
 - 증명은 못할정도로 어렵지만, 구현은 쉽다.
 ![image-20210927120854123](https://raw.githubusercontent.com/is3js/screenshots/main/image-20210927120854123.png)

In [4]:
def collatz(n):
    # 0. n부터 나누기2로 줄거나 or *3+1로 커지는데, 그것을  n==1될때까지 모은다.
    seq = [n]
    # 1. 콜라츠는 1이되면 종료한다. 1보다 클때까지만 돌린다.
    while n>1:
        if n%2==0:
            n=n//2
        else:
            n=3*n + 1
        seq.append(n)
    return seq


collatz(10)

[10, 5, 16, 8, 4, 2, 1]

In [ ]:
for i in range(1, 10):
    print(collatz(i))

### 규칙발견을 위해, seq-len을 end=" "로 출력해보자.
 - 수열의 길이에 규칙이 있는지 봤는데, 안보인다...

In [ ]:
for i in range(1, 100):
    print(len(collatz(i)), end=" ")

### 규칙은 없지만, 쉬워서 문제는 우박수처럼 쏟아진다.

![image-20210927121646260](https://raw.githubusercontent.com/is3js/screenshots/main/image-20210927121646260.png)

- 손으로는 계산못한다.

![image-20210927121732546](https://raw.githubusercontent.com/is3js/screenshots/main/image-20210927121732546.png)



 - 조건을 만족시키는 가장 작은 수는? 
     - **input index가 정해지지 않은 경우, 그냥  while True:로 무한루프  i=0, i+=1로 시작하자.**
     - 1부터 시작하며, +1씩 나가면서, 조건을 만족할 때 바로 return(break)해버리면 `가장 작은 수`다.
        - **두번째 수는? count변수를 만들어 조건 만족할때마다 +1 해서 체크한다.**
     - 끝을 모르는 경우 while True: + break(or return)의 무한루프를 돈다.

In [9]:
# 콜라츠 수열의 길이가 100인 숫자들 중에서 가장 작은 수는?
n=0
while True:
    n+=1
    if len(collatz(n)) == 100:
        break 

print("길이가 100인 가장 작은 우박수는?", n)

길이가 100인 가장 작은 우박수는? 322


In [16]:
# 10000이하의 콜라츠 수열의 길이가 가장긴 수는?
# -> 1부터 10000까지를 돌아야한다 -> for문으로가능 / 너무 많지 않기 때문에 다돌아도 됨.
# -> 돌면서 여러개 중에 가장 긴것 찾기 -> for문위에 변수에 curr변수 update
# --> update 당시의 i도 알아야한다면...  curr = max(curr, ??? )로 한줄업데이트 불가능.ㅠ -> if문으로 풀어서 + 변수도 챙겨야한다.

max_len = 0
max_halestone = []

for n in range(1, 10000+1):
     #curr_len = max(curr_len, len(collatz(i)))
     # 만약, 수열자체도 필요하다면, 
     halestone = collatz(n)
     length = len(halestone)
     if max_len < len(halestone):
         max_len = length
         max_halestone = halestone


print("10,000이하 가장 긴 우박수의 길이와 우박수 :", max_len, max_halestone[0])

10,000이하 가장 긴 우박수의 길이와 우박수 : 262 6171


### 연습문제: 콜라츠수열의 길이가 500보다 큰 첫번째 자연수는?
 - 콜라츠수열의 길이가 500보다 큰 첫번째 자연수는?
     - 첫번째 자연수 -> 조건을 만족하면 바로 return한 수
 - 콜라츠수열의 길이가 500보다 큰 10번째 자연수의  콜라츠 수열 길이는?
     - 10번재 자연수 -> 조건을 만족할 때마다 count따로 해줘야함.

In [22]:
n=0
count = 0
while True:
    n+=1
    if len(collatz(n)) > 500:
        count+=1
        print(count, n)
        if count ==10:
            break

n

1 626331
2 704623


KeyboardInterrupt: 

## 로마 문자(기호) -> 아라비아 숫자로 

![image-20210928165007845](https://raw.githubusercontent.com/is3js/screenshots/main/image-20210928165007845.png)

![image-20210928165014112](https://raw.githubusercontent.com/is3js/screenshots/main/image-20210928165014112.png)



### 로마 숫자 표기법

![image-20210928165213769](https://raw.githubusercontent.com/is3js/screenshots/main/image-20210928165213769.png)

![image-20210928165245431](https://raw.githubusercontent.com/is3js/screenshots/main/image-20210928165245431.png)

- X, XX, XXX : 10, 20, 30
  - **XL : 40**



![image-20210928165353548](https://raw.githubusercontent.com/is3js/screenshots/main/image-20210928165353548.png)

- L, LX, LXX, LXXX : 50, 60, 70, 80
  -  **XC : 90**



![image-20210928165541773](https://raw.githubusercontent.com/is3js/screenshots/main/image-20210928165541773.png)

- C, CC, CCC : 100 200 300 
  - **CD : 400**



![image-20210928165645098](https://raw.githubusercontent.com/is3js/screenshots/main/image-20210928165645098.png)

- D, DC, DCC, DCCC: 500, 600, 700, 800
  - **MC : 900**

![image-20210928165730181](https://raw.githubusercontent.com/is3js/screenshots/main/image-20210928165730181.png)

- 규칙발견하기
  - **기본적으로는 `가산표기법` : 기호의 값을 그냥 합산하면 됨. **
  - **너무 길어지는 경우, `감산표기법`**: 
    - **큰 수 앞에(왼쪽에) 작은 수를 넣으면, 큰 수(오른)에서  작은 수(왼)를 빼라는 뜻.**
    - 예) 4, 9, 40, 90, 400, 900 ..





In [25]:
string = "MDCLXVI"

print(string[0])
print(string[6])

for i in range(len(string)):
    print(string[i], end=" ")
print()

print("M" + "D" + "C" + "L" + "X" + "V" + "I")

M
I
M D C L X V I 
MDCLXVI


### getValueOf(symbol) : 로마 문자를 아라비아 숫자로 변환

In [31]:
# 함수로 매칭..
def getValueOf(symbol):
    if symbol == 'M': return 1000
    elif symbol == 'D': return 500
    elif symbol == 'C': return 100
    elif symbol == 'L': return 50
    elif symbol == 'X': return 10
    elif symbol == 'V': return 5
    elif symbol == 'I': return 1

roman = "MDCLXVI"

for i in range(len(roman)):
    #print(getValueOf(roman[i]))
    print(roman[i] ,":", getValueOf(roman[i]))



M : 1000
D : 500
C : 100
L : 50
X : 10
V : 5
I : 1


### dict로 매칭되는 데이터 매핑
 - 함수로 길게 if elif...로 매칭을 시켜 return해줄 필요가 없다.

In [35]:
romanDict = {
    'M':1000, 'D':500, 'C':100, 'L':50, 'X':10, 'V':5, 'I':1
}

roman = "MDCLXVI"

for i in range(len(roman)):
    #print(getValueOf(roman[i]))
    print(roman[i] ,":", romanDict[roman[i]])

M : 1000
D : 500
C : 100
L : 50
X : 10
V : 5
I : 1


### 단순 가산만 해주기
 - 가산표기법은 고려안한 상태로 로마문자->숫자변환->단순 가산만
 - **어차피 숫자로 변환후 각 대응되는 숫자를 누적해서 더하던지 빼던지 해야함 -> 변수에 누적합**

In [37]:
roman = "MDCLXVI"

arabic = 0

for i in range(len(roman)):
    #arabic += getValueOf(roman[i])
    arabic += romanDict[roman[i]]

print(arabic)


1666


### 감산 생각해주기
 - i번째 수에서, i+1(다음수)보다 작은 숫자라고 판명되면 arabic에 -= 빼주면 된다.
     - for문에 i+1이 들어온 순간,, 바깥  for문에서는 1개 적게 끝나고 + 맨 마지막꺼는 추가로 더해준다?
     - 또는, 여기서는, 감산 확인 if문 자체를... 마지막 인덱스는 포함안되도록, i <( n-1)로 추가 조건주기


In [41]:
def toArabicNumber(roman):
    arabic = 0

    for i in range(len(roman)):
        # 마지막 [n-1]에서는 검사안하도록.. and 조건을 주자. 
        # - 그외에는.. 다 검사하게 되어있다..
        if (i< len(roman)-1) and romanDict[roman[i]] < romanDict[roman[i+1]]:
            arabic -= romanDict[roman[i]]
        else:
            arabic += romanDict[roman[i]]

    return arabic


romans = ['CCLXIX', 'LXXX', 'XXIX']
arabics =[]
for roman in romans:
    arabics.append(toArabicNumber(roman))

print(arabics)

[269, 80, 29]


### 연습문제: 아라비아 숫자를 로마문자로 변환가능할까?
 - 참고 : www.wolframalpha.com
     - `roman numerals 1666`
     - `roman numerals MDCLXVI`